src/zope/sendmail/mailer.py

```diff
@@ -52,19 +64,21 @@ def vote(self, fromaddr, toaddrs, message):
 
         self.code, self.response = code, response
 
-
-    def abort(self):
-        if self.connection is None:
-            return
-
+    def _close_connection(self):
         try:
             self.connection.quit()
         except SSLError:
-            #something weird happened while quiting
+            # something weird happened while quiting
             self.connection.close()
+        self.connection = None
+
+    def abort(self):
+        if self.connection is None:
+            return
+        self._close_connection()
 
     def send(self, fromaddr, toaddrs, message):
-        connection = getattr(self, 'connection', None)
+        connection = self.connection
```

Comment:
Ok, this will not work because thread locals are initialized only for the original thread:

```
>>> import threading
>>> l = threading.local()
>>> l.connection = None
>>> l.connection
>>> threading.Thread(target=lambda: l.connection).start()
>>> Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "<stdin>", line 1, in <lambda>
AttributeError: 'thread._local' object has no attribute 'connection'
```

To fix this change `_make_property()` to use `getattr(self._smtp, name, None)`.

There's also a nicer but more verbose way -- you can subclass `threading.local` and define `connection` et al as class attributes.  You could even create a `__init__()` method and it'll be automatically called for each thread, as described by `pydoc _threading_local`.

Reply:
Those are absolutely good points. I don't know why I skipped over all of that. Thanks for catching this, sorry I wasn't more thorough.

Reply:
That's what reviews are for! ;)